File: bizprint-server-java/src/main/java/com/brainsellers/bizstream/bizprint/PDFDirectPrintStream.java
/*
 * Copyright 2024 BrainSellers.com Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.brainsellers.bizstream.bizprint;

import java.io.IOException;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.logging.Logger;

import javax.servlet.http.HttpServletResponse;

/**
 *  ダイレクト印刷のストリームを生成するためのクラスです。
 */
public class PDFDirectPrintStream extends PDFCommonPrintStream {
    private static final Logger LOGGER = Logger.getLogger(PDFDirectPrintStream.class.getName());

    /**
     *  HTTP 応答オブジェクト
     */
    protected HttpServletResponse response;

    /**
     *  印刷応答 URL
     */
    protected String responseUrl;

    /**
     *  ファイル保存
     */
    protected String saveFileName;

    /**
     *  ブラウザ target 名
     */
    protected String target;

    /**
     *  印刷ダイアログ表示フラグ
     */
    protected Boolean printDialog;

    /**
     *  ダウンロードファイル名の一意化フラグ
     */
    protected Boolean sppnameUnified = true;


    /**
     *  インスタンスを生成し、指定された HTTP 応答オブジェクトで初期化します。
     *
     *      @param      response    HTTP 応答オブジェクト
     */
    public PDFDirectPrintStream(HttpServletResponse response) {
        super();


        this.response = response;

        responseUrl  = null;
        saveFileName = null;
        target       = null;
        printDialog  = null;
    }

    /**
     *  バッファをクローズします。
     */
    public void close() throws IOException {
        pdfdoc.close();

        CreateEncryptSpp cesp = new CreateEncryptSpp();

        StringBuilder tmp = new StringBuilder(parameters());
        String sep = tmp.length() == 0 ? "" : "\n";

        if (responseUrl != null) {
            tmp.append(sep + "responseURL=" + responseUrl);
            sep = "\n";
        }
        if (saveFileName != null) {
            tmp.append(sep + "saveFileName=" + saveFileName);
            sep = "\n";
        }
        if (target != null) {
            tmp.append(sep + "target=" + target);
            sep = "\n";
        }
        if (printDialog != null) {
            tmp.append(sep + "printDialog=" + (String.valueOf(printDialog.booleanValue())));
            sep = "\n";
        }

        tmp.append(sep);

        cesp.setPassword(userPassword);
        byte[] sendspp = null;
        try {
            sendspp = cesp.createCompressedSpp(tmp.toString(), jobName, pdfdoc);
        } catch (Exception e) {
        }
        if (sendspp == null) {
            LOGGER.severe("send spp file is NULL.");
            throw new IOException("send spp file is NULL.");
        }

        response.setContentLength(sendspp.length);
        response.setContentType("application/x-spp");

        //実行ごとにファイル名を変える
        if(sppnameUnified){
            //現在時刻からファイル名作成
            Date nowdate = new Date();
            int rand = (int) (Math.random() * 1000);

            String fname_now = new SimpleDateFormat("yyyyMMdd_HHmmss").format(nowdate) + "_" + String.format("%04d", rand) + ".spp";

            //attachmentだとブラウザによってはダウンロード後に手動ファイル保存になるためinlineにする
            response.addHeader("Content-Disposition", "inline; filename=" + fname_now );
        }

        OutputStream output = getOutput(response.getOutputStream());

        output.write(sendspp);
        output.close();
    }

    /**
     *  印刷応答 URL をセットします。
     *
     *      @param      value       印刷応答 URL
     */
    public void setResponseUrl(String value) {
        responseUrl = value;
    }

    /**
     *  ファイル保存をセットします。
     *
     *      @param      value       ファイル保存
     */
    public void setSaveFileName(String value) {
        saveFileName = value;
    }

    /**
     *  ブラウザ target 名をセットします。
     *
     *      @param      value       ブラウザ target 名
     */
    public void setTarget(String value) {
        target = value;
    }

    /**
     *  印刷ダイアログ表示フラグをセットします。<BR>
     *  <BR>
     *  このメソッドでtrue(印刷ダイアログを表示する)を指定した場合、以下のメソッドでの指定は無視されます。<BR>
     *  <UL>
     *  <LI>{@link PDFCommonPrintStream#setNumberOfCopy PDFCommonPrintStream#setNumberOfCopy(印刷部数)}
     *  <LI>{@link PDFCommonPrintStream#setFromPage PDFCommonPrintStream#setFromPage(開始ページ番号)}
     *  <LI>{@link PDFCommonPrintStream#setToPage PDFCommonPrintStream#setToPage(終了ページ番号)}
     *  <LI>{@link PDFCommonPrintStream#setDoFit PDFCommonPrintStream#setDoFit(ページサイズに合わせて印刷フラグ)}
     *  </UL>
     *
     *      @param      value       印刷ダイアログ表示フラグ
     */
    public void setPrintDialog(boolean value) {
        printDialog = Boolean.valueOf(value);
    }

    /**
     *  SPPファイル名を一意化するかを指定します(デフォルト: true)
     *      @param      value       true: 一意化する, false: 一意化しない
     */
    public void setSppNameUnified(Boolean value) {
        sppnameUnified = value;
    }
}

File: bizprint-server-csharp/src/PDFDirectPrintStream.cs
// Copyright 2024 BrainSellers.com Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Web;

[assembly: CLSCompliant(true)]

namespace com.brainsellers.bizstream.bizprint
{
    /**
     * <summary>
     * ダイレクト印刷のストリームを生成するためのクラスです。
     * </summary>
     */
    public class PDFDirectPrintStream : PDFCommonPrintStream
    {
        /// <summary>HTTP 応答オブジェクト</summary>
        protected HttpResponse response;

        /// <summary>印刷応答 URL</summary>
        protected String responseUrl;

        /// <summary>ファイル保存</summary>
        protected String saveFileName;

        /// <summary>ブラウザ target 名</summary>
        protected String target;

        /// <summary>印刷ダイアログ表示フラグ</summary>
        protected Boolean printDialog;

        /// <summary>sppファイル名の一意化</summary>
        protected Boolean sppnameUnified = true;

        /// <summary>印刷応答 URL</summary>
        protected String userPassword;

        /**
         * <summary>
         * <c>PDFDirectPrintStream</c>クラスの新規インスタンスを生成し初期化します。
         * </summary>
         * <param name="response">HttpResponseオブジェクト</param>
         * <remarks>
         * <c>PDFDirectPrintStream</c>クラスの新規インスタンスを生成し、
         * 指定された HTTP 応答オブジェクトで初期化します。
         * </remarks>
         * <example>
         * <code>
         *    Dim direct As PDFDirectPrintStream
         *    direct = New PDFDirectPrintStream(Response)
         * </code>
         * </example>
         */
        public PDFDirectPrintStream(HttpResponse response)
        {
            this.response = response;
            this.response.Clear();

            responseUrl = null;
            saveFileName = null;
            target = null;
            printDialog = false;
            userPassword = "";
        }

        /**
         * <summary>印刷ダイアログ表示フラグをセットします。</summary>
         * <param name="value">印刷ダイアログ表示フラグ</param>
         * <remarks>
         * <para>このメソッドでtrue(印刷ダイアログを表示する)を指定した場合、以下のメソッドでの指定は無視されます。</para>
         * <para>PDFCommonPrintStream#setNumberOfCopy(印刷部数)</para>
         * <para>PDFCommonPrintStream#setFromPage(開始ページ番号)</para>
         * <para>PDFCommonPrintStream#setToPage(終了ページ番号)</para>
         * <para>PDFCommonPrintStream#setDoFit(ページサイズに合わせて印刷フラグ)</para>
         * </remarks>
         */
        public void setPrintDialog(Boolean value)
        {
            printDialog = value;
        }

        /**
         * <summary>印刷応答 URL をセットします。</summary>
         * <param name="value">印刷応答 URL</param>
         */
        public void setResponseUrl(String value)
        {
            responseUrl = value;
        }

        /**
         * <summary>ファイル保存をセットします。</summary>
         * <param name="value">ファイル保存</param>
         */
        public void setSaveFileName(String value)
        {
            saveFileName = value;
        }

        /**
         * <summary>ブラウザ target 名をセットします。</summary>
         * <param name="value">ブラウザ target 名</param>
         */
        public void setTarget(String value)
        {
            target = value;
        }

        /**
         * <summary>ユーザパスワードをセットします。(平文)</summary>
         * <param name="value">平文パスワード</param>
         */
        public void setPassword(String value)
        {
            if (value == null)
            {
                throw new ArgumentException("setPassword");
            }
            userPassword = value;
        }

        /**
         * <summary>ユーザパスワードをセットします。(base64エンコード済み)</summary>
         * <param name="value">base64エンコード済みパスワード</param>
         */
        public void setPasswordWithEncoded(String value)
        {
            if (value == null)
            {
                //例外Throw
                throw new ArgumentException("setPasswordWithEncoded");
            }

            try
            {
                byte[] decbytes = Convert.FromBase64String(value);
                Encoding enc = Encoding.GetEncoding("UTF-8");
                userPassword = enc.GetString(decbytes);

            }
            catch (Exception ex)
            {
                //例外Throw
                throw new ArgumentException("setPasswordWithEncoded Cannot Decode. " + ex.Message);
            }
        }

        /**
         * <summary>sppファイル名の一意化フラグをセットします。</summary>
         * <param name="value">trueの場合、sppファイル名を一意にします</param>
         */
        public void setSppNameUnified(Boolean value)
        {
            sppnameUnified = value;
        }

        /**
         * <summary>バッファをクローズします。</summary>
         */
        public void Close()
        {
            StringBuilder tmp = new StringBuilder(parameters());
            String sep = (tmp.Length == 0) ? "" : "\n";

            if (responseUrl != null)
            {
                tmp.Append(sep + "responseURL=" + responseUrl);
                sep = "\n";
            }

            if (saveFileName != null)
            {
                tmp.Append(sep + "saveFileName=" + saveFileName);
            }

            if (target != null)
            {
                tmp.Append(sep + "target=" + target);
            }
            tmp.Append(sep + "printDialog=" + printDialog.ToString().ToLower() + sep);


            CreateEncryptSpp createEnc = new CreateEncryptSpp();
            createEnc.setPassword(userPassword);
            byte[] sendspp = createEnc.createCompressedSpp(tmp.ToString(), jobName, pdfdoc.ToArray());

            response.ContentType = "application/x-spp";
            response.AddHeader("Content-Length", sendspp.Length.ToString());

            if (sppnameUnified)
            {
                //現在時刻からファイル名作成
                DateTime DT = new DateTime();
                DT = DateTime.Now;
                Random cRandom = new System.Random();
                int rand = cRandom.Next(1000);
                string fname_now = DT.ToString("_yyMMdd_HHmmss_") + rand.ToString("D4");
                response.AddHeader("Content-Disposition", "inline; filename=" + fname_now+".spp");
            }

            Stream output = response.OutputStream;

            output.Write(sendspp, 0, (int)sendspp.Length);
            output.Close();

            pdfdoc.Close();
        }
    }
}
